Refactor/data_transformers

mabilton opened the pull request:
<!-- Please mention an issue this pull request addresses. -->
Fixes #1407.

### Summary

This PR makes two major changes to how data transformers work:
1. Fixed parameters and fitted parameters (if any) are now 'automatically' passed to `ts_transform`, `ts_inverse_transform`, and `ts_fit` by default, without the user having to re-implement `_transform_iterator`, `_inverse_transform_iterator`, or `_fit_iterator`. To accommodate this, each of these `ts_*` methods now accepts a `params` (dictionary) argument, where `params['fixed']` stores the fixed parameters of the transformation (defined to be all those attributes defined in the child-most class *before calling `super().__init__`*) and `params['fitted']` stores the fitted parameters (i.e. what `ts_fit` returned).
2. `component_mask` key word arguments will be automatically applied to timeseries inputs given to `ts_*` and automatically 'unapplied' to timeseries outputs returned by these methods, which means that users don't have to worry about 'manually' dealing with these arguments inside of their implemented `ts_*` methods.  If the user does not wish for `component_masks` to be automatically applied, they may specify `mask_components=False` when calling `super().__init__`; this will cause any `component_mask` key word argument to be passed via `kwargs` to the called method (i.e. current behaviour).

To see how these changes can help simplify the work involved in implementing a new transformation, compare the [current implementation of `BoxCox`](https://github.com/unit8co/darts/blob/master/darts/dataprocessing/transformers/boxcox.py) and with the `BoxCox` implementation in this PR. 

### Other Information

Some other minor changes that come with this PR:
1. I split up the `_reshape_in` method into two new methods (`apply_component_mask` and `stack_samples`), as well as `_reshape_out` into two new methods (`unapply_component_mask` and `unstack_samples`). There are two reasons for this change:
     - `_reshape_in` and `_reshape_out` were responsible for applying two distinctly different changes to the data: masking component columns and stacking the samples of each component along a single axis. From a user interaction and maintainability perspective, I think it's much cleaner to have these two pieces of functionality separated from one another. The names `_reshape_in` and `_reshape_out` are, in my opinion, also a bit vague.
     - In the original implementation of `_reshape_in`/`_reshape_out`, the 'stacking step' was performed using a `for` loop; I've changed this so that only `np.swapaxes` and `np.reshape` operations are used , which theoretically should speed things up a bit.
     - I've made these new methods explicitly public, so that it's explicitly clear to new users looking to write their own transformations that it's "okay" to call these functions. 
2. I've refactored all of the currently implemented transformations so that they conform with these changes, with the exception of `StaticCovariatesTransformer`, which directly overrides the `fit`, `inverse_transform` and `transform` methods anyways. Similarly, I also had to make some minor adjustments to existing tests.
3. I've implemented new tests for the refactored `BaseDataTransformer` , `FittableDataTransformer`, and `InvertableDataTransformer` classes.
4. Some transformations, such as `BoxCox`, allow for different fixed parameter values to be distributed over different parallel jobs. To facillitate this, I've added a `parallel_params` argument to the `*Transformer` classes, which allows the user to specify which parameters should take different values for different parallel jobs. 

There are two drawbacks to what I've done here:
1. Some of these changes are slightly breaking; with that being said, I doubt many users currently have their own private code bases with their own custom-implemented `darts` transformations (although I could be wrong).
2. One 'gotcha' with this proposed interface is  that the user must only call `super().__init__` *after initialising the fixed parameters of their transformation*. For example, the following will allow the user to access `'_my_param'` in `params['fixed']`:
   ```python
   class MyTransform(BaseDataTransformer):
       def __init__(self):
              # Correct: Define fixed parameter *before* calling `super().__init__`
              self._my_param = 1
              super().__init__()
   ```
   whereas the following will **not**:
   ```python
   class MyTransform(BaseDataTransformer):
       def __init__(self):
              # Incorrect: Define fixed parameter *after* calling `super().__init__`
              super().__init__()
              self._my_param = 1
   ```

Any thoughts/comments on these changes are more than welcome.

Cheers,
Matt.

mabilton commented:
So I see that the following line in cell `[40]` within the `00-quickstart.ipynb` is throwing an error:
```python
# scale back:
pred_air = scaler.inverse_transform(pred_air)
```
From what I can see, `scaler` is fitted to two different timeseries in cell `[31]`:
```python
scaler = Scaler()
train_air_scaled, train_milk_scaled = scaler.fit_transform([train_air, train_milk])
```
What's the intended behaviour for a data transformation which is trained on two series but is given only a single series to inverse transform? I was clearly under the misunderstanding that an error should be thrown in such cases.

hrzn commented:
Thanks a lot @mabilton ! We just need a bit more time to get to it and review :)

madtoinou commented on darts/dataprocessing/transformers/diff.py:
```diff
@@ -226,14 +210,18 @@ def ts_inverse_transform(
+                "Cannot specify `component_mask` with `dropna = True`, "
+                "since differenced and undifferenced components will be "
+                "of different lengths."
+            ),
```
Maybe reference the logger?

madtoinou commented on darts/dataprocessing/transformers/fittable_data_transformer.py:
```diff
@@ -50,14 +69,20 @@ def __init__(
         can easily introduce a bottleneck and nullify parallelisation benefits.
         """
-        super().__init__(name=name, n_jobs=n_jobs, verbose=verbose)
+        super().__init__(
```
shouldn't the `super().__init__()` be called after instantiating the other attributes or is it valid only for the transforms that directly inherit from `BaseDataTransformer`?

hrzn replied:
My understanding is that this is necessary in the childmost classes only (where the instance variables are defined).

madtoinou reviewed (commented):
LGTM, I wrote some minor comments about details.

Thank you for this PR that should solve many of the issues related to data transformers!

mabilton commented:
Hey @madtoinou - thanks for all the useful feedback. Just letting you know that I'm pretty busy in my personal life at the moment, so it'll probably take a day or two to implement your suggestions. Apologies for the delay.

Cheers,
Matt.

hrzn commented on darts/dataprocessing/transformers/base_data_transformer.py:
```diff
@@ -81,7 +127,7 @@ def set_n_jobs(self, value: int):
     @staticmethod
     @abstractmethod
-    def ts_transform(series: TimeSeries) -> TimeSeries:
+    def ts_transform(series: TimeSeries, params: Mapping[str, Any]) -> TimeSeries:
```
You could add `params` to the docstring.

hrzn commented on darts/dataprocessing/transformers/base_data_transformer.py:
```diff
@@ -184,115 +242,238 @@ def transform(
+            vals = vals.all_values()
+        shape = vals.shape
+        new_shape = (shape[0] * shape[2], shape[1])
+        stacked = np.swapaxes(vals, 1, 2).reshape(new_shape)
```
That's better indeed :)

hrzn commented on darts/dataprocessing/transformers/window_transformer.py:
```diff
@@ -88,49 +78,38 @@ def __init__(
             Specifies how to treat missing values that were added by the window transformations
             at the beginning of the resulting TimeSeries. By default, Darts will leave NaNs in the resulting TimeSeries.
             This parameter can be one of the following:
-
```
I think removing these blank lines before a list in rst format might have adverse effect once this is rendered in HTML. Could you perhaps check rendering the doc in HTML (`./gradlew buildDocs`) or - if you want to be on the safe side - simply keep these blank lines.

hrzn commented on darts/tests/dataprocessing/transformers/test_fittable_data_transformer.py:
```diff
@@ -0,0 +1,272 @@
+from darts.utils.timeseries_generation import constant_timeseries
+
+
+class BaseDataTransformerTestCase(unittest.TestCase):
```
Nice unit tests, thanks!

hrzn reviewed (approved):
That looks _really_ good to me. We can almost merge it as such IMO :) Only got a few pretty minor comments. Nice job @mabilton! And sorry for the super late review...

mabilton commented:
Hey @hrzn , @madtoinou.

I've (finally) found some time to implement your suggestions in addition to the following changes:
1. I've added more detail to the docstrings of the transformer classes/methods.
2. I've added a 'global_fit' option to the `FittableDataTransformer`. When `global_fit = True`, the entire `Sequence` of `TimeSeries` passed to the `fit` method will be passed to `ts_fit`, thereby allowing the user to fit a single set of parameters to all of the (disjoint) `TimeSeries` provided by the user, as opposed to independently fitting a different set of parameters to each `TimeSeries`. At this point in time, only the `BoxCox`, `Scaler`, and `StaticCovariatesTransformer` (can) use this option. Indeed, my motivation for introducing this `global_fit` option was so I could refactor the `StaticCovariatesTransformer`, which I'll talk about now.
3. In the last iteration of this PR, the `StaticCovariatesTransformer` was basically left untouched and was implemented in a way that did not utilise the refactored `fit`, `transform`, and `inverse_transform` methods. Since fitting the `StaticCovariatesTransformer` requires first collating the static covariates of all of the provided `TimeSeries`, I had to first had introduce a 'global fitting' option before refactoring `StaticCovariatesTransformer` (as mentioned in point 2). With that done, I could then refactor `StaticCovariatesTransformer` so that only a `ts_transform`, `ts_inverse_transform`, and a `ts_fit` method needed to be implemented (as opposed to directly overriding the `transform`, `inverse_transform`, and `fit` methods).

As an aside, it appears that `statsforecast==1.5.0` (which was only released a few days ago) introduces some dependency conflicts, and resulted in some tests failing. It looks like this is a problem on the `statsforecast` side, so to get around this for the time being, I've specified `statsforecast>=1.4,<1.5` in `requirements/core.txt`.

Hopefully that all makes sense, and please let me know what you guys think. Thanks in advance for any help.

Cheers,
Matt.

hrzn reviewed (approved):
LGTM, nice job @mabilton !
